File: two_pointers/two_sum_2/cpp/two_sum_2.cpp
#include<iostream>
#include<vector>

using std::cout;
using std::vector;

// Time O(n)
// Space O(1)

class Solution {
public:
    static vector<int> twoSum(vector<int>& numbers, int target) {
        int l = 0;
        int r = numbers.size()-1;

        while(l < r) {
            int sum = numbers[l] + numbers[r];
            if(sum < target) {
                l++;
            } else if(sum > target) {
                r--;
            } else {
                return { l+1, r+1 };
            }
        }
        return {};
    }
};

int main() {
    vector<int> numbers = { 2, 7, 11, 15 };
    int target = 9;

    auto answer = Solution::twoSum(numbers, target);

    for(auto& n : answer) {
        cout << n << ' ';
    }
    cout << '\n';
}

File: two_pointers/two_sum_2/js/two_sum_2.js
const numbers = [ 2, 7, 11, 15 ];
const target = 9;

const twoSum = (numbers, target) => {
  let l = 0;
  let r = numbers.length-1;

  while(l < r) {
    const sum = numbers[l] + numbers[r];
    if(sum < target) {
      l++;
    } else if(sum > target) {
      r--;
    } else {
      return [l+1, r+1];
    }
  }
  return [];
}

console.log(twoSum(numbers, target));
